src/psyclone/tests/psyir/nodes/reference_test.py

```diff
@@ -214,3 +214,221 @@ def test_reference_can_be_copied():
     # Modifying the new reference does not affect the original
     ref1._symbol = scalar_symbol
     assert ref.symbol is array_symbol
+
+
+def test_reference_next_access(fortran_reader):
+    '''Test the next_access function for a Reference'''
+    code = '''subroutine my_sub()
+    integer :: a
+    integer :: b
+    a = 1
+    b = 1
+    a = b
+    end subroutine my_sub'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    b = routine.children[1].lhs
+    a_2 = routine.children[2].lhs
+    b_2 = routine.children[2].rhs
+    assert a.next_access() is a_2
+    assert b.next_access() is b_2
+    assert a_2.next_access() is None
+
+    code = '''subroutine my_sub()
+    integer :: a
+    integer :: b
+    a = 1
+    do a = 0, 10
+        b = a
+    end do
+    end subroutine'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    loop = routine.children[1]
+    b_a = loop.loop_body.children[0].lhs
+    a_2 = loop.loop_body.children[0].rhs
+    assert a.next_access() is loop
+    assert b_a.next_access() is None
+
+    # Check the function for basic structures
+    code = '''subroutine my_sub()
+    type :: x
+       integer :: a
+       real :: b
+    end type
+    type(x) :: i
+    i%a = 2
+    i%b = 0.1
+    i%a = 3
+    i%b = 0.2
+    end subroutine'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    b = routine.children[1].lhs
+    a_2 = routine.children[2].lhs
+    b_2 = routine.children[3].lhs
+    assert a.next_access() is a_2
+    assert b.next_access() is b_2
+    assert a_2.next_access() is None
+    assert b_2.next_access() is None
+
+    # Check the function for array access
+    code = '''subroutine my_sub()
+    integer, dimension(100) :: a
+    a(0) = 2
+    a(1) = 2
+    end subroutine my_sub'''
+    psyir = fortran_reader.psyir_from_source(code)
+    routine = psyir.children[0]
+    a = routine.children[0].lhs
+    a_2 = routine.children[1].lhs
+    assert a.next_access() is a_2
+    assert a_2.next_access() is None
+
+    # Check if statements don't affect
+    code = '''subroutine my_sub()
+   integer :: a
+   logical :: b = .false.
+   a = 1
+   if(b) then
+     a = 0
+   end if
+   a = 2
```

Comment:
I am curious about what happens with an else block:
```
a = 1
if b then
  a = 2
else
  a = 3
endif
a = 4
```

Does next_access of a(2) points to a(3) or a(4)?

I don't know what is the right behaviour, we can refine that on upcoming PRs that actually use this functionality, but it would be good for now to have an assert here to demonstrate what it currently does.

Reply:
It will point to a = 3.
I'm not sure I agree with that behaviour but I'm also not sure whether we should change it here not - perhaps more of a discussion when we create thing using this code.